File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_typescript-node/src/rules.ts
import type { Combatant, ScoredCombatant } from './types.js';

export const CR_XP: Record<string, number> = {
  '0': 10,
  '1/8': 25,
  '1/4': 50,
  '1/2': 100,
  '1': 200,
  '2': 450,
  '3': 700,
  '4': 1100,
  '5': 1800,
};

export const LEVEL_3_THRESHOLDS = { easy: 75, medium: 150, hard: 225, deadly: 400 };

export const ENCOUNTER_THRESHOLDS: Record<number, { easy: number; medium: number; hard: number; deadly: number }> = {
  1: { easy: 25, medium: 50, hard: 75, deadly: 100 },
  2: { easy: 50, medium: 100, hard: 150, deadly: 200 },
  3: { easy: 75, medium: 150, hard: 225, deadly: 400 },
  4: { easy: 125, medium: 250, hard: 375, deadly: 500 },
  5: { easy: 250, medium: 500, hard: 750, deadly: 1100 },
  6: { easy: 300, medium: 600, hard: 900, deadly: 1400 },
  7: { easy: 350, medium: 750, hard: 1100, deadly: 1700 },
  8: { easy: 450, medium: 900, hard: 1400, deadly: 2100 },
  9: { easy: 500, medium: 1100, hard: 1700, deadly: 2600 },
  10: { easy: 600, medium: 1200, hard: 1900, deadly: 2800 },
  11: { easy: 800, medium: 1600, hard: 2400, deadly: 3600 },
  12: { easy: 1000, medium: 2000, hard: 3000, deadly: 4500 },
  13: { easy: 1100, medium: 2200, hard: 3400, deadly: 5100 },
  14: { easy: 1250, medium: 2500, hard: 3800, deadly: 5700 },
  15: { easy: 1400, medium: 2800, hard: 4300, deadly: 6400 },
  16: { easy: 1600, medium: 3200, hard: 4800, deadly: 7200 },
  17: { easy: 2000, medium: 3900, hard: 5900, deadly: 8800 },
  18: { easy: 2100, medium: 4200, hard: 6300, deadly: 9500 },
  19: { easy: 2400, medium: 4900, hard: 7300, deadly: 10900 },
  20: { easy: 2800, medium: 5700, hard: 8500, deadly: 12700 },
};

/** Parse a dice expression of the form `<count>d<sides>[+-<mod>]`. */
export function parseDiceExpression(expression: unknown): { diceCount: number; sides: number; modifier: number } | null {
  if (typeof expression !== 'string') return null;
  const match = expression.trim().match(/^([0-9]+)d([0-9]+)(?:([+-])([0-9]+))?$/);
  if (!match) return null;
  const diceCount = Number(match[1]);
  const sides = Number(match[2]);
  const modifier = match[3] ? (match[3] === '+' ? 1 : -1) * Number(match[4]) : 0;
  if (diceCount <= 0 || sides <= 0) return null;
  return { diceCount, sides, modifier };
}

/** Standard D&D ability modifier: floor((score - 10) / 2). */
export function abilityModifier(score: number): number {
  return Math.floor((score - 10) / 2);
}

/** Standard 5e proficiency bonus by character level, capped at 6. */
export function proficiencyBonus(level: number): number {
  if (level <= 4) return 2;
  if (level <= 8) return 3;
  if (level <= 12) return 4;
  if (level <= 16) return 5;
  return 6;
}

const CLASS_HIT_DICE: Record<string, number> = {
  barbarian: 12,
  bard: 8,
  cleric: 8,
  druid: 8,
  fighter: 10,
  monk: 8,
  paladin: 10,
  ranger: 10,
  rogue: 8,
  sorcerer: 6,
  warlock: 8,
  wizard: 6,
};

const HIT_DIE_AVERAGE: Record<number, number> = {
  6: 4,
  8: 5,
  10: 6,
  12: 7,
};

/** Return the starting hit die for a valid class, or null for unknown classes. */
export function classHitDie(className: string): number | null {
  return CLASS_HIT_DICE[className.toLowerCase()] ?? null;
}

/** Return the fixed average roll for a known hit die size. */
export function averageHitDie(sides: number): number | null {
  return HIT_DIE_AVERAGE[sides] ?? null;
}

/** Encounter multiplier based on the number of enemy monsters. */
export function multiplierForMonsterCount(count: number): number {
  if (count === 1) return 1;
  if (count === 2) return 1.5;
  if (count <= 6) return 2;
  if (count <= 10) return 2.5;
  if (count <= 14) return 3;
  return 4;
}

export function difficultyFromAdjustedXP(
  adjusted: number,
  thresholds: { easy: number; medium: number; hard: number; deadly: number },
): string {
  if (adjusted >= thresholds.deadly) return 'deadly';
  if (adjusted >= thresholds.hard) return 'hard';
  if (adjusted >= thresholds.medium) return 'medium';
  if (adjusted >= thresholds.easy) return 'easy';
  return 'trivial';
}

export function recommendationForDifficulty(difficulty: string): string {
  switch (difficulty) {
    case 'trivial':
      return 'no challenge';
    case 'easy':
      return 'safe warm-up';
    case 'medium':
      return 'solid challenge';
    case 'hard':
      return 'tough fight';
    case 'deadly':
      return 'deadly threat';
    default:
      return 'unknown';
  }
}

/** Produces a fallback open-thread title from a recap summary. */
export function resolveTrailThread(summary: string, fallbackMonster: string): string {
  const match = summary.match(/the (\S+) trail/i);
  const monster = match ? match[1] : fallbackMonster;
  return `Resolve ${monster} trail ambush`;
}

/** Compute a deterministic initiative score from the raw roll and Dexterity. */
export function scoreCombatant(c: Combatant): ScoredCombatant {
  return { ...c, score: c.roll + c.dex };
}

/**
 * Sort combatants by initiative score descending, then Dexterity descending,
 * then name ascending for deterministic tie-breaking.
 */
export function sortCombatants(combatants: Combatant[]): ScoredCombatant[] {
  return combatants
    .map(scoreCombatant)
    .sort((a, b) => {
      if (b.score !== a.score) return b.score - a.score;
      if (b.dex !== a.dex) return b.dex - a.dex;
      return a.name.localeCompare(b.name);
    });
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T052124Z_claude_claude-sonnet-5_python-stdlib/rules.py
"""D&D 5e rule tables and pure calculation helpers.

Everything here is stateless: constants describing rule-book data plus
functions that transform request-shaped input into rule-derived output.
No I/O, no HTTP, no persistence. `server.py` handlers call into this module
and only concern themselves with request parsing / response shaping.
"""
import hashlib
import hmac
import re
import secrets

# Base XP awarded per monster, keyed by challenge rating (string, since "1/4"
# etc. are not valid dict/JSON numeric keys in the source rule tables).
CR_XP = {
    "0": 10,
    "1/8": 25,
    "1/4": 50,
    "1/2": 100,
    "1": 200,
    "2": 450,
    "3": 700,
    "4": 1100,
    "5": 1800,
}

# Encounter difficulty XP thresholds per party-member level. Only level 3 is
# populated; other levels intentionally fall through to "unsupported level".
LEVEL_THRESHOLDS = {
    3: {"easy": 75, "medium": 150, "hard": 225, "deadly": 400},
}

# Matches dice expressions like "2d6", "1d20+3", "4d8-1".
DICE_RE = re.compile(r"^(\d+)d(\d+)([+-]\d+)?$")

# Wizard spell slots by character level; only level 5 is populated.
WIZARD_SPELL_SLOTS = {
    5: {"1": 4, "2": 3, "3": 2},
}

# Maximum prepared spells per class, keyed by character level; only wizard
# level 1 is populated (max 1 prepared spell).
MAX_PREPARED_SPELLS = {
    "wizard": {1: 1},
}

# Spell slots available for *casting* (as opposed to the PHB spell-slots
# reference table above), keyed by class -> character level -> spell level.
# Cantrips (spell level 0) are an at-will resource in 5e and are represented
# with a large sentinel count rather than a strict slot pool. Only wizard
# level 1 is populated, matching MAX_PREPARED_SPELLS.
CASTING_SPELL_SLOTS = {
    "wizard": {1: {0: 9999, 1: 1}},
}

DIFFICULTY_RECOMMENDATIONS = {
    "trivial": "no real threat",
    "easy": "safe warm-up",
    "medium": "balanced challenge",
    "hard": "tough fight, expect resource drain",
    "deadly": "deadly encounter, prepare escape options",
}

# DM loot-parcel presets by tier; only tier 1 is populated.
DM_LOOT_TIERS = {
    1: {"coins_gp": 75, "items": [{"slug": "healing-potion", "quantity": 2}]},
}

# (level_low, level_high, proficiency_bonus)
PROFICIENCY_TABLE = [
    (1, 4, 2),
    (5, 8, 3),
    (9, 12, 4),
    (13, 16, 5),
    (17, 20, 6),
]

# PHB skill -> governing ability, keyed by skill slug.
SKILL_ABILITY = {
    "acrobatics": "dex",
    "animal-handling": "wis",
    "arcana": "int",
    "athletics": "str",
    "deception": "cha",
    "history": "int",
    "insight": "wis",
    "intimidation": "cha",
    "investigation": "int",
    "medicine": "wis",
    "nature": "int",
    "perception": "wis",
    "performance": "cha",
    "persuasion": "cha",
    "religion": "int",
    "sleight-of-hand": "dex",
    "stealth": "dex",
    "survival": "wis",
}

USERNAME_RE = re.compile(r"^[a-z0-9_-]{2,32}$")
SLUG_RE = re.compile(r"^[a-z0-9-]{1,64}$")

# Level-1 hit die base per class (PHB). hp_max at level 1 = base + con modifier.
CLASS_HIT_DIE = {
    "barbarian": 12,
    "fighter": 10,
    "paladin": 10,
    "ranger": 10,
    "bard": 8,
    "cleric": 8,
    "druid": 8,
    "monk": 8,
    "rogue": 8,
    "warlock": 8,
    "sorcerer": 6,
    "wizard": 6,
}

VALID_RACES = {
    "human", "elf", "dwarf", "halfling", "dragonborn",
    "gnome", "half-elf", "half-orc", "tiefling",
}

# Known spells per class, keyed by spell_id -> (canonical name, level).
# Only "wizard" is populated: wizards may know any spell on their list;
# classes absent from this table (e.g. rogue) may not learn any spells.
CLASS_SPELL_LIST = {
    "wizard": {
        "fire-bolt": ("Fire Bolt", 0),
        "mage-hand": ("Mage Hand", 0),
        "prestidigitation": ("Prestidigitation", 0),
        "ray-of-frost": ("Ray of Frost", 0),
        "magic-missile": ("Magic Missile", 1),
        "shield": ("Shield", 1),
        "detect-magic": ("Detect Magic", 1),
        "identify": ("Identify", 1),
    },
}

VALID_BACKGROUNDS = {
    "acolyte", "criminal", "folk-hero", "noble", "sage", "soldier",
    "charlatan", "entertainer", "guild-artisan", "hermit", "outlander",
    "sailor",
}

VALID_INVENTORY_ITEM_IDS = {
    "healing-potion", "torch",
    "leather-armor", "ring-of-protection", "amulet-of-health",
}

VALID_EQUIPMENT_SLOTS = {"armor", "accessory"}

ITEM_EQUIPMENT_SLOT = {
    "leather-armor": "armor",
    "ring-of-protection": "accessory",
    "amulet-of-health": "accessory",
}

ATTUNABLE_ITEM_IDS = {"ring-of-protection", "amulet-of-health"}

MAX_ATTUNEMENTS = 1

CONSUMABLE_ITEM_EFFECTS = {
    "healing-potion": {"type": "healing", "hp_restored": 5},
}


def max_prepared_spells(klass, level):
    return MAX_PREPARED_SPELLS.get(klass, {}).get(level, 0)


def casting_spell_slots(klass, level, spell_level):
    """Total slots of `spell_level` available to `klass` at `level` (0 if none)."""
    return CASTING_SPELL_SLOTS.get(klass, {}).get(level, {}).get(spell_level, 0)


def level_one_hp_max(klass, con_modifier):
    return max(1, CLASS_HIT_DIE[klass] + con_modifier)


def hp_gain_per_level(klass, con_modifier):
    """Deterministic (fixed, non-random) HP gain for levels beyond 1.

    Uses the standard 5e "fixed value" in place of rolling the hit die:
    half the die size rounded up, plus one (e.g. 1d8 -> 5).
    """
    die = CLASS_HIT_DIE[klass]
    return max(1, die // 2 + 1 + con_modifier)


def is_plain_int(x):
    """True for real ints, excluding bool (bool is a subclass of int)."""
    return isinstance(x, int) and not isinstance(x, bool)


def ability_modifier(score):
    return (score - 10) // 2


SEASON_OFFSETS = {"spring": 0, "summer": 1, "autumn": 2, "winter": 3}
WEATHER_BY_OFFSET = {0: "clear", 1: "rain", 2: "wind", 3: "snow"}


def weather_for(day, season):
    offset = (day + SEASON_OFFSETS[season]) % 4
    return WEATHER_BY_OFFSET[offset]


def proficiency_bonus(level):
    for lo, hi, bonus in PROFICIENCY_TABLE:
        if lo <= level <= hi:
            return bonus
    return None


def multiplier_for_count(n):
    """Encounter-multiplier table from the DMG, keyed by monster count."""
    if n <= 1:
        return 1
    if n == 2:
        return 1.5
    if 3 <= n <= 6:
        return 2
    if 7 <= n <= 10:
        return 2.5
    if 11 <= n <= 14:
        return 3
    return 4


def hash_password(password, salt=None):
    if salt is None:
        salt = secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), bytes.fromhex(salt), 200_000)
    return salt, digest.hex()


def verify_password(password, salt, expected_hex):
    _, digest_hex = hash_password(password, salt)
    return hmac.compare_digest(digest_hex, expected_hex)
